tests/proxytest.py

```diff
@@ -356,3 +356,36 @@ async def test_websockets(proxy, launch_backend):
         port = await websocket.recv()
 
     assert port == str(default_backend_port)
+
+
+async def test_autohttps(autohttps_toml_proxy, pebble, launch_backend):
+    proxy = autohttps_toml_proxy
+
+    routespec = "/autohttps"
+    target = "http://127.0.0.1:9900"
+
+    backend_port = 9900
+    launch_backend(backend_port)
+
+    await wait_for_services([proxy.public_url, target])
+
+    await proxy.add_route(routespec, target, {})
+
+    # Test the actual routing
+    req = HTTPRequest(proxy.public_url + routespec, method="GET", validate_cert=False)
+    resp = await AsyncHTTPClient().fetch(req)
+    backend_response = int(resp.body.decode("utf-8"))
+
+    # Test we were redirected to https
+    # https://127.0.0.1:8443/autohttps
+    expected_final_redirect_url = (
+        "https://"
+        + urlparse(proxy.public_url).hostname
+        + ":"
+        + str(proxy.traefik_https_port)
+        + routespec
+    )
+    assert resp.effective_url == expected_final_redirect_url
+
+    # Test redirection to the route added
+    assert backend_response == backend_port
```

Comment:
Are we only testing redirection in this test?

It would perhaps make sense to test a successful response to https as well perhaps, than proves that TLS termination (encrypt/decrypt etc) is done successfully. The client that makes this webrequest would need to not try verify the TLS certificate (`verify=False` in python's requests library, as passed with the `get`, `post`, and etc functions).

An issue I experienced with z2jh was that I had to await traefik being ready, and found that awaiting the logs from Pebble saying it has issued a certificate was good enough for my test suite to ensure my traefik instance was ready, while it wasn't enough for me to await the log entry "Registering certificate..." made by traefik - i think 2-5 seconds additional to that entry showing was needed for traefiks HTTPS endpoints to be ready.

Reply:
> Are we only testing redirection in this test?

The test should follow this flow:
* launch a [http server on 9900](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L368) (that only knows to respond to GET request with the port is running on).
* [wait for traefik and the http server](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L370) show up on their specified ports
* [add a route to traefik](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L372) from `/autohttps` to the backend http server
* make [a http request](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L372) for the public_url and the route we just added.
The request should be redirected as:
`http://localhost:8000/autohttps` -> (because we're redirecting traffic from the http entrypoint on port 8000 to https) 
`https://localhost:8443/autohttps` -> (because we added the route to the backend dummy server)
`http://localhost:9900` that should respond with a 200 code and `9900` as the payload.

I thought that testing this redirection will mean that we also got a successful response to https :confused: 

> The client that makes this webrequest would need to not try verify the TLS certificate (verify=False in python's requests library, as passed with the get, post, and etc functions)

The request has the validate_cert flag set to False [here](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/tests/proxytest.py#L375).

> An issue I experienced with z2jh was that I had to await traefik being ready, and found that awaiting the logs from Pebble saying it has issued a certificate was good enough for my test suite to ensure my traefik instance was ready, while it wasn't enough for me to await the log entry "Registering certificate..." made by traefik - i think 2-5 seconds additional to that entry showing was needed for traefiks HTTPS endpoints to be ready.

I noticed that registering the certificate takes a while. From the logs, it looks that the request seems to happen after the certificate acquisition. I'm wondering if it's just out of luck or [this check](https://github.com/GeorgianaElena/traefik-proxy/blob/auto_https/test/proxytest.py#L370) is enough or maybe the certificate acquisition is blocking (I didn't find any evidence of this in traefik's code to support this though).

Reply:
It looks like the test is indeed not testing the right thing: https://travis-ci.org/github/GeorgianaElena/traefik-proxy/jobs/688791781 passes even though it cannot connect to the acme client :confused: